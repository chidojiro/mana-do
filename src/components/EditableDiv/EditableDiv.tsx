import React from 'react';

interface Props {
  content: string;
  onComplete: (value: string) => void;
}

export const EditableDiv: React.FC<Props> = ({ content, onComplete }) => {
  const [isInEditMode, setIsInEditMode] = React.useState(false);

  const [value, setValue] = React.useState(content);

  const cancelEdit = () => {
    setValue(content);
    setIsInEditMode(false);
  };

  const handleEditComplete = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'Enter') {
      onComplete(value);
      setIsInEditMode(false);
    }

    if (event.key === 'Escape') {
      cancelEdit();
    }
  };

  return (
    <div>
      {!isInEditMode ? (
        <div onDoubleClick={() => setIsInEditMode(true)}>{content}</div>
      ) : (
        <input
          autoFocus
          value={value}
          onKeyDown={handleEditComplete}
          onChange={(e) => setValue(e.target.value)}
          onBlur={cancelEdit}
        />
      )}
    </div>
  );
};
