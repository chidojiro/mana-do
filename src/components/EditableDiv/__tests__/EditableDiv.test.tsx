import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

import { EditableDiv, Props } from '../EditableDiv';

describe('Switch', () => {
  const defaultProps: Props = { content: 'content example', onComplete: jest.fn() };

  const enterEditMode = (wrapper: ShallowWrapper) => {
    const content = wrapper.find('.content--test');
    content.simulate('doubleClick');
  };

  const requestComplete = (wrapper: ShallowWrapper, newContent: string) => {
    enterEditMode(wrapper);

    wrapper.find('input').simulate('change', { target: { value: newContent } });
    wrapper.find('input').simulate('keyDown', { key: 'Enter' });
  };

  const testContentAfterCancel = (wrapper: ShallowWrapper) => {
    expect(wrapper.find('.content--test').exists()).toBe(true);

    enterEditMode(wrapper);

    expect(wrapper.find('input').prop('value')).toBe(defaultProps.content);
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render the component', () => {
    const wrapper = shallow(<EditableDiv {...defaultProps} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should initially not in edit mode', () => {
    const wrapper = shallow(<EditableDiv {...defaultProps} />);

    const content = wrapper.find('.content--test');

    expect(content.exists()).toBe(true);
  });

  it('should enter edit mode when content is double clicked', () => {
    const wrapper = shallow(<EditableDiv {...defaultProps} />);

    enterEditMode(wrapper);

    const input = wrapper.find('input');

    expect(input.exists()).toBe(true);
  });

  it('should call onComplete with new content', () => {
    const wrapper = shallow(<EditableDiv {...defaultProps} />);

    const newContent = 'new content';

    requestComplete(wrapper, newContent);

    expect(defaultProps.onComplete).toBeCalledWith(newContent);
  });

  it('should not call onComplete and stay editing when the new content is empty', () => {
    const wrapper = shallow(<EditableDiv {...defaultProps} />);

    const newContent = '';

    requestComplete(wrapper, newContent);

    expect(defaultProps.onComplete).not.toBeCalled();
  });

  it('should not call onComplete and stay editing when the new content is the same', () => {
    const wrapper = shallow(<EditableDiv {...defaultProps} />);

    const newContent = defaultProps.content;

    requestComplete(wrapper, newContent);

    expect(defaultProps.onComplete).not.toBeCalled();
  });

  it('should cancel edit and discard changes when Escape is pressed', () => {
    const wrapper = shallow(<EditableDiv {...defaultProps} />);

    enterEditMode(wrapper);

    wrapper.find('input').simulate('change', { target: { value: 'new content' } });
    wrapper.find('input').simulate('keyDown', { key: 'Escape' });

    testContentAfterCancel(wrapper);
  });

  it('should cancel edit and discard changes when the input is blurred', () => {
    const wrapper = shallow(<EditableDiv {...defaultProps} />);

    enterEditMode(wrapper);

    wrapper.find('input').simulate('blur');

    testContentAfterCancel(wrapper);
  });
});
