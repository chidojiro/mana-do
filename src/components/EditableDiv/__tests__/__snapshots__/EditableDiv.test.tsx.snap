// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render the component 1`] = `
<div>
  <div
    className="content--test"
    onDoubleClick={[Function]}
  >
    content example
  </div>
</div>
`;
